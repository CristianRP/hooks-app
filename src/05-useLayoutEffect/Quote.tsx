import { useLayoutEffect, useRef, useState } from 'react';

export type Quote = {
  author: string;
  quote: string;
}

type BoxSize = {
  width: number;
  height: number;
}

export const Quote = ({ author, quote }: Quote) => {

  const pRef = useRef<HTMLParagraphElement>(null);
  const [boxSize, setBoxSize] = useState<BoxSize>({ width: 0, height: 0})

  useLayoutEffect(() => {
    if (pRef.current) {
      const { height, width } = pRef.current.getBoundingClientRect();
      setBoxSize({height, width});
    }
  }, [quote]);

  return (
    <>
      <figure className="max-w-screen-md mx-auto">
        <svg className="w-10 h-10 mx-auto mb-3 text-gray-400 dark:text-gray-600" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 18 14">
            <path d="M6 0H2a2 2 0 0 0-2 2v4a2 2 0 0 0 2 2h4v1a3 3 0 0 1-3 3H2a1 1 0 0 0 0 2h1a5.006 5.006 0 0 0 5-5V2a2 2 0 0 0-2-2Zm10 0h-4a2 2 0 0 0-2 2v4a2 2 0 0 0 2 2h4v1a3 3 0 0 1-3 3h-1a1 1 0 0 0 0 2h1a5.006 5.006 0 0 0 5-5V2a2 2 0 0 0-2-2Z"/>
        </svg>
        <blockquote className='flex flex-col'>
          <p ref={ pRef } className='mt-4 text-2xl self-end'>
            { quote }
          </p>
          <footer className='self-end text-xl'>
            { author }
          </footer>
        </blockquote>
        <figcaption className="flex items-center justify-center mt-6 space-x-3 rtl:space-x-reverse">
            <img className="w-6 h-6 rounded-full" src="https://flowbite.s3.amazonaws.com/blocks/marketing-ui/avatars/michael-gouch.png" alt="profile picture" />
            <div className="flex items-center divide-x-2 rtl:divide-x-reverse divide-gray-500 dark:divide-gray-700">
                <cite className="pe-3 font-medium text-gray-900">{ author }</cite>
                <cite className="ps-3 text-sm text-gray-500 dark:text-gray-400">CEO at Google</cite>
            </div>
        </figcaption>
        <code>
          { JSON.stringify(boxSize) }
        </code>
      </figure>
    </>
  )
}

