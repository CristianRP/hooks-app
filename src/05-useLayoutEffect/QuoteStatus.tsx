export const QuoteStatus = ({ message = 'Loading', color }: { message: string, color?: string }) => {
  return (
    <div className={`p-4 bg-sky-400 rounded-md text-white w-400 ${color}`}>
      { message }
    </div>
  )
}
