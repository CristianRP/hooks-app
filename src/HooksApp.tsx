export const HooksApp = () => {
  return (
    <div className="container">
      <h1 className="text-3xl font-bold underline hover:text-red-200">
        Hooks App!
      </h1>
    </div>
  )
}

