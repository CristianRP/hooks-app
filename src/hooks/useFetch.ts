import { useEffect, useState } from 'react'

export type DataType<T> = {
  [K in keyof T]: T[K] extends (string | object | number | T[K])[] ? (string | object | number | T[K])[] : T[K];
}

type FetchResponse<T> = {
  data: DataType<T>;
  isLoading: boolean;
  hasError: boolean;
  error?: {
    code: number;
    message: string;
  }
}

export const useFetch = <T extends object>( url: string ): FetchResponse<T> => {
  const [state, setState] = useState({
    data: {},
    isLoading: true,
    hasError: false,
    error: {},
  });

  useEffect(() => {

    const getFetch = async() => {
      const resp = await fetch(url);


      console.log('parent-resp', resp);
      
      if ( !resp.ok ) {
        setState({
          data: {},
          isLoading: false,
          hasError: true,
          error: {
            code: resp.status,
            message: resp.statusText
          }
        });
        return;
      }

      const data = await resp.json();
      console.log(data);
      

      setState({
        data,
        isLoading: false,
        hasError: false,
        error: {}
      });
    }

    getFetch();
  }, [url]);

  return {
    data: state.data as DataType<T>,
    isLoading: state.isLoading,
    hasError: state.hasError,
  }
}
