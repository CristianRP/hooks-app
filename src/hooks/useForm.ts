import { ChangeEvent, useState } from 'react';

type FormState<T> = {
  [K in keyof T]: string;
}

type UseForm<T> = {
  formState: FormState<T>;
  onInputChange: (event: ChangeEvent<HTMLInputElement>) => void;
  onResetForm: () => void;
} & {
  [K in keyof FormState<T>]: string;
}

export const useForm = <T extends object>( initialForm: T = {} as T): UseForm<T> => {
  const [formState, setFormState] = useState<FormState<T>>( initialForm as FormState<T>);


  const onInputChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = target;
    setFormState({
      ...formState,
      [ name ]: value
    });
  }

  const onResetForm = () => {
    setFormState(initialForm as FormState<T>)
  }

  return {
    formState,
    onInputChange,
    onResetForm,
    ...formState,
  }
}
