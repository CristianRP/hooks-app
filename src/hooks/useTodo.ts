import { useEffect, useReducer } from 'react';
import { Todo } from '../08-useReducer/TodoList';
import { Action, todoReducer } from '../08-useReducer/todoReducer';

type DispatchAction = (action: Action) => void;

const initialState: Todo[] = []

const init = () => {
  return JSON.parse( localStorage.getItem('todos')! ) || [];
}

export const useTodo = () => {

  const [ todos, dispatch ]:[ Todo[], DispatchAction ] = useReducer( todoReducer , initialState, init );

  const handleNewTodo = ( todo: Todo ) => {
    dispatch({
      type: '[TODO] Add Todo',
      payload: todo
    })
  }

  const handleDeleteTodo = ( id:string ) => {
    dispatch({
      type: '[TODO] Remove Todo',
      payload: {
        id
      }
    })
  }

  const handleToggleTodo = ( id:string ) => {
    dispatch({
      type: '[TODO] Update Done Todo',
      payload: {
        id
      }
    })
  }

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos || []));
  },[todos]);

  return {
    todos,
    handleNewTodo,
    handleDeleteTodo,
    handleToggleTodo,
  }
}

