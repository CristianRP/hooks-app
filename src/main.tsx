import React from 'react'
import ReactDOM from 'react-dom/client'


import './index.css'
import { SimpleFormWithHook } from './02-useEffect/SimpleFormWithHook'

ReactDOM.createRoot(document.getElementById('root')!).render(
  // <React.StrictMode>
    <SimpleFormWithHook />
  // </React.StrictMode>,
)
