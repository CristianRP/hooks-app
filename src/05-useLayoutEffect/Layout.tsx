import { useFetch } from '../hooks'
import { QuoteStatus } from './QuoteStatus'
import { Quote } from './Quote'
import { useState } from 'react'

export const Layout = () => {

  const [ reload, setReload ] = useState(false);
  const { isLoading, hasError, data } = useFetch<Quote[]>(`https://api.breakingbadquotes.xyz/v1/quotes/`, false, reload);
  
  return (
    <div className='flex flex-col'>
      <h1 className='text-3xl font-bold'>BreakingBad Quotes</h1>
      <hr className='mt-2' />


      {
        !isLoading && hasError && <QuoteStatus message='Error' color='bg-red-600' />
      }

      {
        isLoading && !hasError && <QuoteStatus message='Loading...' />
      }

      {
        !isLoading && <Quote {...data[0]} />
      }

      <button
        className='m-2 p-2 bg-green-800 rounded-md text-white active:bg-green-700 hover:bg-green-600 focus:outline-none focus:ring focus:ring-green-500'
        onClick={ () => setReload(!reload) }
        >
        Next Quote
      </button>
    </div>
  )
}
