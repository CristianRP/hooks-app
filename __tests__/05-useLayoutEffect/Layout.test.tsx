import React from 'react';
import { render, screen } from '@testing-library/react';

import { Layout } from '../../src/05-useLayoutEffect/Layout';
import { useFetch } from '../../src/hooks';

jest.mock('../../src/hooks/useFetch');

describe('Tests on <Layout /> Componet', () => {

  beforeAll(() => {
    jest.mocked(useFetch).mockReturnValue({
      data: {},
      isLoading: true,
      hasError: false
    })
  })

  test('should match snapshot', () => {
    const { container } = render(<Layout />);

    expect( container ).toMatchSnapshot();
  });
  
  test('should show the component by default', () => {
    render(<Layout />);

    expect( screen.getByText('Loading...') );
    expect( screen.getByText('BreakingBad Quotes') );

    const nextButton = screen.getByRole<HTMLButtonElement>('button', { name: 'Next Quote' });

    console.log(nextButton.disabled);
    
    expect( nextButton.disabled ).toBeTruthy();
  });

  test('should show a Quote', () => {
    jest.mocked(useFetch).mockReturnValue({
      data: [{
        author: 'Cristian',
        quote: 'Test Quote'
      }],
      isLoading: false,
      hasError: false
    })

    render(<Layout />);

    expect( screen.getByText('Test Quote') ).toBeTruthy();
    expect( screen.getAllByText('Cristian') ).toBeTruthy();

    const nextButton = screen.getByRole<HTMLButtonElement>('button', { name: 'Next Quote' });
    expect( nextButton.disabled ).toBeFalsy();

  });

});