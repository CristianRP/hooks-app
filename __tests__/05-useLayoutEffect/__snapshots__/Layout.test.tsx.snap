// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <Layout /> Componet should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <h1
      class="text-3xl font-bold"
    >
      BreakingBad Quotes
    </h1>
    <hr
      class="mt-2"
    />
    <div
      class="p-4 bg-sky-400 rounded-md text-white w-400 undefined"
    >
      Loading...
    </div>
    <button
      class="m-2 p-2 bg-green-800 rounded-md text-white active:bg-green-700 hover:bg-green-600 focus:outline-none focus:ring focus:ring-green-500"
      disabled=""
    >
      Next Quote
    </button>
  </div>
</div>
`;
